pandas/core/internals/managers.py

```diff
@@ -572,7 +572,12 @@ def setitem(self, indexer, value) -> Self:
                         0, blk_loc, values
                     )
                     # first block equals values
-                    self.blocks[0].setitem((indexer[0], np.arange(len(blk_loc))), value)
+                    col_indexer: slice | np.ndarray
+                    if isinstance(indexer[1], slice) and indexer[1] == slice(None):
+                        col_indexer = slice(None)
+                    else:
+                        col_indexer = np.arange(len(blk_loc))
+                    self.blocks[0].setitem((indexer[0], col_indexer), value)
```

Comment:
It's not clear to me why it's correct to pass through `slice(None)` but not other cases, e.g. `slice(0, 3, 2)` or `[0, 1]`.

Reply:
I think you're right. My current changes failed under the following test case. I'll mark the PR as a draft until I properly handle the bug and will also add more test cases. Thanks!

```python
dftest3 = pd.DataFrame(
    {"A": [1, 4, 1, 5], "B": [2, 5, 2, 6], "C": [3, 6, 1, 7], "D": [8, 9, 10, 11]}
)
df3 = dftest3[["B", "C", "D"]]
df3.iloc[[1, 3], 0:3:2] = [[2, 2], [2, 2]]
```